tests/mocks/core/services/request/RequestMock.php

```diff
@@ -0,0 +1,211 @@
+        $this->ip_address = $this->visitorIp($ip_address);
+        parent::__construct($get, $post, $cookie, $server);
+    }
+
+
+    /**
+     * @param string $key
+     * @param string $value
+     * @param bool   $override_ee
+     * @return    void
+     */
+    public function setRequestParam($key, $value, $override_ee = false)
+    {
+        $this->request[ $key ] = $value;
+    }
+
+
+    /**
+     * returns the value for a request param if the given key exists
+     *
+     * @param string $key
+     * @param null   $default
+     * @return mixed
+     */
+    public function getRequestParam($key, $default = null)
+    {
+        return isset($this->request[ $key ]) ? $this->request[ $key ] : $default;
+    }
+
+
+    /**
+     * check if param exists
+     *
+     * @param string $key
+     * @return bool
+     */
+    public function requestParamIsSet($key)
+    {
+        return isset($this->request[ $key ]);
+    }
+
+
+    /**
+     * check if a request parameter exists whose key that matches the supplied wildcard pattern
+     * and return the value for the first match found
+     * wildcards can be either of the following:
+     *      ? to represent a single character of any type
+     *      * to represent one or more characters of any type
+     *
+     * @param string     $pattern
+     * @param null|mixed $default
+     * @return false|int
+     */
+    public function getMatch($pattern, $default = null)
+    {
+        // replace wildcard chars with regex chars
+        $pattern = str_replace(
+            array("\*", "\?"),
+            array('.*', '.'),
+            preg_quote($pattern, '/')
+        );
+        foreach ($this->request as $key => $request_param) {
+            if (preg_match('/^' . $pattern . '$/is', $key)) {
+                return $request_param;
+            }
+        }
+        return null;
+    }
+
+
+    /**
+     * check if a request parameter exists whose key matches the supplied wildcard pattern
+     * wildcards can be either of the following:
+     *      ? to represent a single character of any type
+     *      * to represent one or more characters of any type
+     * returns true if a match is found or false if not
+     *
+     * @param string $pattern
+     * @return false|int
+     */
+    public function matches($pattern)
+    {
+        return $this->getMatch($pattern) !== null;
+    }
+
+
+    /**
+     * remove param
+     *
+     * @param string $key
+     * @param bool   $unset_from_global_too
+     */
+    public function unSetRequestParam($key, $unset_from_global_too = false)
+    {
+        unset($this->request[ $key ]);
+    }
+
+
+    /**
+     * @return string
+     */
+    public function visitorIp($visitor_ip = '0.0.0.0')
+    {
+        return $visitor_ip;
+    }
+
+
+    /**
+     * @return string
+     */
+    public function requestUri()
+    {
+        return isset($this->server['REQUEST_URI']) ? $this->server['REQUEST_URI'] : '';
```

Comment:
Is this method in production code covered in tests anywhere?  The logic here is significantly different than that in `core/services/request/Request`

Reply:
Again, the `RequestMock` class is NOT for testing the `Request` class.
It appears there is no test for `requestUri()` in `EventEspresso\tests\testcases\core\request_stack\RequestTest` but that has no bearing on what is in the `RequestMock` class.